packages/internals/src/cli/getSchema.ts

```diff
@@ -16,38 +29,84 @@ export async function getSchemaPath(
   opts: { cwd: string } = {
     cwd: process.cwd(),
   },
-): Promise<string | null> {
+): Promise<GetSchemaResult | null> {
   return getSchemaPathInternal(schemaPathFromArgs, {
     cwd: opts.cwd,
   })
 }
 
+async function readSchemaFromSingleFile(schemaPath: string): Promise<GetSchemaResult> {
+  const file = await readFile(schemaPath, { encoding: 'utf-8' })
+  return { schemaPath, files: file } as const
+}
+
+async function readSchemaFromMultiFiles(schemaPath: string): Promise<GetSchemaResult | null> {
+  const files = await loadSchemaFiles(schemaPath)
+
+  // TODO: problem: if the Prisma config isn't valid, we currently get a
+  // `Error: Could not find a schema.prisma file that is required for this command.` error
+  // in the multi-file case.
+  const config = await getConfig({
+    datamodel: files,
+    ignoreEnvVarErrors: true,
+  })
+  const previewFeatures = config.generators.find((g) => g.previewFeatures.length > 0)?.previewFeatures
+  const usesPrismaSchemaFolder = (previewFeatures || []).includes('prismaSchemaFolder')
+
+  if (usesPrismaSchemaFolder) {
+    return { schemaPath, files } as const
+  }
+
+  return null
+}
+
 export async function getSchemaPathInternal(
   schemaPathFromArgs?: string,
   opts: { cwd: string } = {
     cwd: process.cwd(),
   },
-): Promise<string | null> {
-  if (schemaPathFromArgs) {
-    // 1. try the user custom path
-    const customSchemaPath = await getAbsoluteSchemaPath(path.resolve(schemaPathFromArgs))
-    if (!customSchemaPath) {
-      throw new Error(`Provided --schema at ${schemaPathFromArgs} doesn't exist.`)
+): Promise<GetSchemaResult | null> {
+  const trials = (
+    [
+      // 1. try the user custom path
+      {
+        strategy: 'absolute',
+        fn: getAbsoluteSchemaPath,
+        sourcePath: schemaPathFromArgs ? path.resolve(schemaPathFromArgs) : null,
+      },
+
+      // 2. Try the package.json `prisma.schema` custom path
+      { strategy: 'package.json', fn: getSchemaPathFromPackageJson, sourcePath: opts.cwd },
+
+      // 3. Try the conventional ./schema.prisma or ./prisma/schema.prisma paths
+      { strategy: 'relative', fn: getRelativeSchemaPath, sourcePath: opts.cwd },
+
+      // 4. Try resolving yarn workspaces and looking for a schema.prisma file there
+      { strategy: 'yarn', fn: resolveYarnSchema, sourcePath: opts.cwd },
+    ] as const
+  ).filter(({ sourcePath }) => sourcePath !== null)
+
+  for (const { strategy, sourcePath, fn } of trials) {
+    debug(`Trying ${strategy}...`)
+
+    const schemaPath = await fn(sourcePath!)
+    debug(`${strategy} resolved to ${schemaPath}`)
+
+    if (!schemaPath) {
+      continue
     }
 
-    return customSchemaPath
-  }
+    const schemaPathResult = await Promise.any([
```

Comment:
Right now, this is undefined behavoir if both file and folder exist: whoever completes first wins and gets used.
Instead we need check if both exist and throw if that's the case.

Reply:
Correct: this should run `readSchemaFromMultiFiles` and fall back to `readSchemaFromSingleFile`.